src/Microsoft.AspNet.FileProviders.Combined/CombinedFileProvider.cs

```diff
@@ -0,0 +1,168 @@
+        }
+
+        /// <summary>
+        /// Enumerate a directory at the given path, if any.
+        /// </summary>
+        /// <param name="subpath">The path that identifies the directory</param>
+        /// <returns>Contents of the directory. Caller must check Exists property. The content is a merge of the contents of the provided <see cref="IFileProvider"/>. When there is multiple <see cref="IFileInfo"/> with the same Name property, only the first one is included on the results.</returns>
+        public IDirectoryContents GetDirectoryContents(string subpath)
+        {
+            // gets the content of all directories and merged them
+            var existingDirectoryContentsForAllProviders = _fileProviders.Select(
+                fileProvider => fileProvider.GetDirectoryContents(subpath))
+                .Where(directoryContents => directoryContents != null && directoryContents.Exists)
+                .ToList();
+
+            // There is no existing directory contents
+            if (existingDirectoryContentsForAllProviders.Count == 0)
+            {
+                return new NotFoundDirectoryContents();
+            }
+            var combinedDirectoryContents = new CombinedDirectoryContents(existingDirectoryContentsForAllProviders);
+            return combinedDirectoryContents;
+        }
+
+        /// <summary>
+        /// Creates a <see cref="IChangeToken"/> for the specified <paramref name="filter"/>.
+        /// </summary>
+        /// <remarks></remarks>
+        /// <param name="pattern">Filter string used to determine what files or folders to monitor. Example: **/*.cs, *.*, subFolder/**/*.cshtml.</param>
+        /// <returns>An <see cref="IChangeToken"/> that is notified when a file matching <paramref name="filter"/> is added, modified or deleted. The change token will be notified when one of the change token returned by the provided <see cref="IFileProvider"/> will be notified.</returns>
+        public IChangeToken Watch(string pattern)
+        {
+            // Watch all file providers
+            var activeChangeTokens = _fileProviders
+                .Select(fileProvider => fileProvider.Watch(pattern))
+                .Where(changeToken => changeToken != null && changeToken.ActiveChangeCallbacks).ToList();
+
+            // There is no change token with active change callbacks
+            if (activeChangeTokens.Count == 0)
+            {
+                return NoopChangeToken.Singleton;
+            }
+            var combinedFileChangeToken = new CombinedFileChangeToken(activeChangeTokens);
+            return combinedFileChangeToken;
+        }
+
+        private class CombinedDirectoryContents : IDirectoryContents
+        {
+            private readonly Dictionary<string, IFileInfo> _files = new Dictionary<string, IFileInfo>();
+
+            public CombinedDirectoryContents(IEnumerable<IEnumerable<IFileInfo>> listOfFiles)
+            {
+                foreach (var files in listOfFiles)
```

Comment:
I have made the assumption that the directory contents will be pre-calculated (based on the two other providers), but you're right. I can move it to a `Lazy<IEnumerable<IFileInfo>>` (to avoiding dealing with the thread-safe and evaluating only once issues).